staging/vhost-device-gpu/src/main.rs

```diff
@@ -0,0 +1,121 @@
+// VIRTIO GPU Emulation via vhost-user
+//
+// Copyright 2024 Red Hat Inc
+//
+// SPDX-License-Identifier: Apache-2.0 or BSD-3-Clause
+
+use std::{path::PathBuf, process::exit};
+
+use clap::Parser;
+use log::{error, info};
+use thiserror::Error as ThisError;
+use vhost_device_gpu::{
+    device::{self, VhostUserGpuBackend},
+    GpuConfig, GpuMode,
+};
+use vhost_user_backend::VhostUserDaemon;
+use vm_memory::{GuestMemoryAtomic, GuestMemoryMmap};
+
+type Result<T> = std::result::Result<T, Error>;
+
+#[derive(Debug, ThisError)]
+pub(crate) enum Error {
+    #[error("Could not create backend: {0}")]
+    CouldNotCreateBackend(device::Error),
+    #[error("Could not create daemon: {0}")]
+    CouldNotCreateDaemon(vhost_user_backend::Error),
+    #[error("Fatal error: {0}")]
+    ServeFailed(vhost_user_backend::Error),
+}
+
+#[derive(Parser, Debug)]
+#[clap(author, version, about, long_about = None)]
+struct GpuArgs {
+    /// vhost-user Unix domain socket.
+    #[clap(short, long, value_name = "SOCKET")]
+    socket_path: PathBuf,
+    #[clap(short, long, value_enum)]
+    gpu_mode: GpuMode,
+}
+
+impl From<GpuArgs> for GpuConfig {
+    fn from(args: GpuArgs) -> Self {
+        let socket_path = args.socket_path;
+        let gpu_mode: GpuMode = args.gpu_mode;
+
+        GpuConfig::new(socket_path, gpu_mode)
+    }
+}
+
+fn start_backend(config: GpuConfig) -> Result<()> {
```

Comment:
Any reason why is this not a method of GpuConfig?

Reply:
I don't think there is a technical reason. From a coding style point of view it feels kind of weird to me, having the config "object" start the backend, instead of having a function that takes the config and the config is just data. Also `GpuConfig` isn't a builder pattern, where I see how this would be a more common thing to do.

That being said we can definitely change it, just trying to explain the rationale.

Reply:
Yep, feel free to mark the comment as resolved.